integration/update_test.go

```diff
@@ -460,3 +461,211 @@ func TestUpdateSet(t *testing.T) {
+				"For example: {$mod: {<field>: ...}} not {$setOnInsert: string}",
+		},
+		"ErrNil": {
+			filter:      bson.D{{"_id", "nil"}},
+			setOnInsert: nil,
+			err: &mongo.WriteError{
+				Code: 9,
+				Message: "Modifiers operate on fields but we found type null instead. " +
+					"For example: {$mod: {<field>: ...}} not {$setOnInsert: null}",
+			},
+			alt: "Modifiers operate on fields but we found type null instead. " +
+				"For example: {$mod: {<field>: ...}} not {$setOnInsert: null}",
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctx, collection := setup(t)
+
+			opts := options.Update().SetUpsert(true)
+			var res *mongo.UpdateResult
+			res, err := collection.UpdateOne(ctx, tc.filter, bson.D{{"$setOnInsert", tc.setOnInsert}}, opts)
+			if tc.err != nil {
+				require.Nil(t, tc.res)
+				AssertEqualAltWriteError(t, *tc.err, tc.alt, err)
+				return
+			}
+
+			require.NoError(t, err)
+			id := res.UpsertedID
+			assert.NotEmpty(t, id)
+			res.UpsertedID = nil
+
+			expectedRes := notModified
+			if tc.stat != nil {
+				expectedRes = tc.stat
+			}
+			assert.Equal(t, expectedRes, res)
+
+			var actual bson.D
+			err = collection.FindOne(ctx, tc.filter).Decode(&actual)
+			require.NoError(t, err)
+			AssertEqualDocuments(t, tc.res, actual)
+		})
+	}
+}
+
+func TestUpdateMany(t *testing.T) {
+	t.Parallel()
+
+	notModified := &mongo.UpdateResult{
+		MatchedCount:  0,
+		ModifiedCount: 0,
+		UpsertedCount: 1,
+	}
+
+	for name, tc := range map[string]struct {
+		filter bson.D
+		query  bson.D
+		stat   *mongo.UpdateResult
+		res    bson.D
+		err    *mongo.WriteError
+	}{
+		"SetSetOnInsert": {
+			filter: bson.D{{"_id", "test"}},
+			query: bson.D{
+				{"$set", bson.D{{"foo", int32(12)}}},
+				{"$setOnInsert", bson.D{{"value", math.NaN()}}},
+			},
+			res: bson.D{{"_id", "test"}, {"foo", int32(12)}, {"value", math.NaN()}},
+		},
+		"SetIncSetOnInsert": {
+			filter: bson.D{{"_id", "test"}},
+			query: bson.D{
+				{"$set", bson.D{{"foo", int32(12)}}},
+				{"$inc", bson.D{{"foo", int32(1)}}},
+				{"$setOnInsert", bson.D{{"value", math.NaN()}}},
+			},
+			err: &mongo.WriteError{
+				Code:    40,
+				Message: "Updating the path 'foo' would create a conflict at 'foo'",
+			},
+		},
+		"UnknownOperator": {
+			filter: bson.D{{"_id", "test"}},
+			query:  bson.D{{"$foo", bson.D{{"foo", int32(1)}}}},
+			err: &mongo.WriteError{
+				Code:    9,
+				Message: "Unknown modifier: $foo. Expected a valid update modifier or pipeline-style update specified as an array",
+			},
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			ctx, collection := setup(t)
+
+			_, err := collection.InsertMany(ctx, []any{
+				bson.D{{"_id", "string"}, {"value", "foo"}},
+			})
+			require.NoError(t, err)
+
+			opts := options.Update().SetUpsert(true)
+			var res *mongo.UpdateResult
+			res, err = collection.UpdateOne(ctx, tc.filter, tc.query, opts)
```

Comment:
That's what I'm talking above. That's not the query, that's the update part of request.
```go
func (coll *Collection) UpdateOne(ctx context.Context, filter interface{}, update interface{},
	opts ...*options.UpdateOptions) (*UpdateResult, error)
```

Reply:
updated

Reply:
Thanks!